### PR TITLE
Speed up signal.lfilter and add a convolution path for FIR filters

### PR DESCRIPTION
**Pull request by ewmoore**
`signal.lfilter` was doing a large number of unnecessary, repetitious divisions in the main filter loop.  Doing them once at the beginning gives about a 3x speed up.  I'm willing to concede that there might be cases where we don't want to normalize the filter coefficients, but those cases were broken before too.  I'd say that should be a future improvement. Its also not clear to me that this is such a problem for doubles.

As I suggested in #3454, when `a = [1]` we can recognize that we are using an FIR filter and do somewhat less work.  I've tried to make the interface and the errors produced as close as possible between the two paths.  

This sort of goes on top of #4614, but the changes are orthogonal to those so I made a separate PR.

Filtering 1,000,000 points with a 51 tap FIR filter drops from 237 ms to 56 ms for me.  A 51 tap IIR filter drops to 83 ms.  These improvements hold up for much smaller signal lengths as well.

I guess we should add some `lfilter` benchmarks?

I haven't looked at the complex filter loops, there might be something to do there too.

### CODE REVIEW

**Comment by ewmoore** [comment]
Based on my tests, `lfilter` is now as fast as Matlab's filter function.

**Comment by ewmoore** [comment]
They also clearly make this FIR optimization based on the run times matching almost exactly for both FIR and IIR.

**Comment by rgommers** [comment]
Nice speedup. We should try to get this merged soonish. 

Needs a rebase.

**Comment by rgommers** [comment]
Adding a benchmark would indeed be useful.

File: scipy/signal/signaltools.py [hunk 1]
```diff
@@ -873,10 +873,52 @@ def lfilter(b, a, x, axis=-1, zi=None):
     """
     if isscalar(a):
         a = [a]
-    if zi is None:
-        return sigtools._linear_filter(b, a, x, axis)
+    if a[0] == 1 and len(a) == 1:
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: The docstring promises that if ``a[0]`` is not 1, then ``a`` and ``b`` are normalized by ``a[0]``. So probably this path should be used for all length-1 ``a``?

File: scipy/signal/signaltools.py [hunk 2]
```diff
@@ -873,10 +873,52 @@ def lfilter(b, a, x, axis=-1, zi=None):
...
+        if zi is not None:
+            # _linear_filter does not broadcast zi
+            zi = np.asarray(zi)
+            if zi.ndim != x.ndim:
+                raise ValueError('object of too small depth for desired array')
+            expected_shape = list(x.shape)
+            expected_shape[axis] = max(a.shape[0], b.shape[0]) - 1
+            expected_shape = tuple(expected_shape)
+            if zi.shape != expected_shape:
+                raise ValueError('Unexpected shape for zi: expected %s, found %s.' % (expected_shape, zi.shape))
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: this line needs breaking

File: scipy/signal/signaltools.py [hunk 3]
```diff
@@ -873,10 +873,52 @@ def lfilter(b, a, x, axis=-1, zi=None):
...
+            raise ValueError('object of too small depth for desired array')
+        x = np.asarray(x)
+        inputs = [b, a, x]
+        if zi is not None:
+            # _linear_filter does not broadcast zi
+            zi = np.asarray(zi)
+            if zi.ndim != x.ndim:
+                raise ValueError('object of too small depth for desired array')
+            expected_shape = list(x.shape)
+            expected_shape[axis] = max(a.shape[0], b.shape[0]) - 1
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: ``a.shape[0]`` is always 1, so never larger then ``b.shape[0]`` --> can be simplified.

Comment (rgommers) [inline scipy/signal/signaltools.py]: and is there an issue if a and b are both size 1?

File: scipy/signal/signaltools.py [hunk 4]
```diff
@@ -873,10 +873,52 @@ def lfilter(b, a, x, axis=-1, zi=None):
...
+        dtype = np.result_type(*inputs)
+
+        if dtype.char not in 'fdgFDGO':
+            raise NotImplementedError("input type '%s' not supported" % dtype)
+
+        b = np.array(b, dtype=dtype, copy=False)
+        x = np.array(x, dtype=dtype, copy=False)
+
+        out_full = np.apply_along_axis(lambda y: np.convolve(b, y), axis, x)
+        ind = out_full.ndim*[slice(None)]
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: PEP8 nitpick: spaces around ``*``

File: scipy/signal/signaltools.py [hunk 5]
```diff
@@ -873,10 +873,52 @@ def lfilter(b, a, x, axis=-1, zi=None):
...
+        ind[axis] = slice(out_full.shape[axis]-len(b)+1)
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: more pep8 spaces

File: scipy/signal/signaltools.py [hunk 6]
```diff
@@ -873,10 +873,52 @@ def lfilter(b, a, x, axis=-1, zi=None):
...
+        out = out_full[ind]
+        if zi is None:
+            return out
+        else:
+            ind[axis] = slice(out_full.shape[axis]-len(b)+1, None)
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: and here spaces too

File: scipy/signal/signaltools.py (hunk 1 above)
Comment (ewmoore) [inline scipy/signal/signaltools.py]: Good point, any `len(a) == 1` can use this path, even if `a[0] != 1`.

File: scipy/signal/signaltools.py (hunk 3 above)
Comment (ewmoore) [inline scipy/signal/signaltools.py]: No issues.  `lfilter` currently requires `zi` to be `None` if `max(len(b), len(a)) == 1` and this does as well.

**Comment by ewmoore** [comment]
Updated, comments addressed.  I haven't added a benchmark though.

**Comment by rgommers** [comment]
OK, this looks good and is indeed faster also for inputs of O(1k-10k) - so time to merge it. Benchmark (and example for ``lfilter``) can be added later.

**Comment by rgommers** [comment]
Thanks @ewmoore

**Pull request merged**